### PR TITLE
c8d/prune: Fix images being deleted when they're still used with a different reference

### LINKED ISSUE #45732: [containerd snapshotter] Prune attempts to remove untagged/dangling images still in use by running containers
### Description

Running Docker Desktop with the experimental containerd integration for Engine enabled.

After running a prune, I started getting some "Unable to calculate image disk size" errors, and noticed that `docker system df` was returning errors:

```
Error response from daemon: failed to retrieve container list: content digest sha256:<digest>: not found
```

The listed digest corresponding to a running container's image's digest. The image was untagged, though it previously HAD been. Looking back in terminal history, the listed digest was deleted by the prune operation.

### Reproduce

1. `docker run --rm -d nginx`
2. `docker image rm -f nginx`
3. `docker system prune -a`
4. `docker system df`

```
Error response from daemon: failed to retrieve container list: content digest sha256:0bb91b50c42bc6677acff40ea0f050b655c5c2cc1311e783097a04061191340b: not found
```

### Expected behavior

Running `docker system prune -a` does not remove images for running containers:

```
WARNING! This will remove:
  - all images without at least one container associated to them
```

(These images have at least one container associated with them.)

### docker version

```bash
Client:
 Cloud integration: v1.0.33
 Version:           24.0.2
 API version:       1.43
 Go version:        go1.20.4
 Git commit:        cb74dfc
 Built:             Thu May 25 21:51:16 2023
 OS/Arch:           darwin/arm64
 Context:           desktop-linux

Server: Docker Desktop 4.21.0 (111569)
 Engine:
  Version:          24.0.2-38-g8e70a1b23e
  API version:      1.43 (minimum version 1.12)
  Go version:       go1.20.4
  Git commit:       8e70a1b23e965d86ec8c2feb77605196ae124630
  Built:            Fri Jun  2 15:58:50 2023
  OS/Arch:          linux/arm64
  Experimental:     false
 containerd:
  Version:          1.6.21
  GitCommit:        3dce8eb055cbb6872793272b4f20ed16117344f8
 runc:
  Version:          1.1.7
  GitCommit:        v1.1.7-0-g860f061
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Client:
 Version:    24.0.2
 Context:    desktop-linux
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.10.5
    Path:     /Users/milas/.docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.18.1
    Path:     /Users/milas/.docker/cli-plugins/docker-compose
  dev: Docker Dev Environments (Docker Inc.)
    Version:  v0.1.0
    Path:     /Users/milas/.docker/cli-plugins/docker-dev
  extension: Manages Docker extensions (Docker Inc.)
    Version:  v0.2.19
    Path:     /Users/milas/.docker/cli-plugins/docker-extension
  init: Creates Docker-related starter files for your project (Docker Inc.)
    Version:  v0.1.0-beta.4
    Path:     /Users/milas/.docker/cli-plugins/docker-init
  sbom: View the packaged-based Software Bill Of Materials (SBOM) for an image (Anchore Inc.)
    Version:  0.6.0
    Path:     /Users/milas/.docker/cli-plugins/docker-sbom
  scan: Docker Scan (Docker Inc.)
    Version:  v0.26.0
    Path:     /Users/milas/.docker/cli-plugins/docker-scan
  scout: Command line tool for Docker Scout (Docker Inc.)
    Version:  v0.12.0
    Path:     /Users/milas/.docker/cli-plugins/docker-scout

Server:
 Containers: 2
  Running: 2
  Paused: 0
  Stopped: 0
 Images: 0
 Server Version: 24.0.2-38-g8e70a1b23e
 Storage Driver: stargz
  driver-type: io.containerd.snapshotter.v1
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 3dce8eb055cbb6872793272b4f20ed16117344f8
 runc version: v1.1.7-0-g860f061
 init version: de40ad0
 Security Options:
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 5.15.49-linuxkit-pr
 Operating System: Docker Desktop
 OSType: linux
 Architecture: aarch64
 CPUs: 5
 Total Memory: 7.667GiB
 Name: docker-desktop
 ID: 56ef67c8-804a-4b75-b1d0-16afaa01050d
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: http.docker.internal:3128
 HTTPS Proxy: http.docker.internal:3128
 No Proxy: hubproxy.docker.internal
 Experimental: false
 Insecure Registries:
  hubproxy.docker.internal:5555
  127.0.0.0/8
 Live Restore Enabled: false
```

### Additional Info

I'm a Docker Inc employee / running an internal/nightly build of Docker Desktop

### PR DESCRIPTION
**Pull request by vvoland**
- Fixes: https://github.com/moby/moby/issues/45732

**- What I did**
Fixed `image prune -a` deleting dangling images used by containers.

**- How I did it**
-  Exclude dangling tag of the images used by containers (this makes sure that if the original image was deleted we don't delete the dangling image)
-  Exclude images that are used by containers which were started by specifying an image ID instead of a specific name.

**- How to verify it**
```bash
$ make TEST_INTEGRATION_USE_SNAPSHOTTER=1 \
       DOCKER_GRAPHDRIVER=overlayfs \
       TEST_FILTER=TestPruneDontDeleteUsedDangling \
       test-integration
...
INFO: Testing against a local daemon
=== RUN   TestPruneDontDeleteUsedDangling
--- PASS: TestPruneDontDeleteUsedDangling (0.07s)
PASS

DONE 1 tests in 1.079s
```

**- Description for the changelog**
```release-notes
containerd integration: `docker image prune -a` no longer removes images still in use by running containers
```

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Review by laurazard — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
Looks like there's a failure;

```
=== FAIL: github.com/docker/docker/integration/image TestPruneDontDeleteUsedDangling (0.03s)
    prune_test.go:23: assertion failed: error is not nil: Error response from daemon: No such image: sha256:0df1207206e5288f4a989a2f13d1f5b3c4e70467702c1d5d21dfc9f002b7bd43
```

**Comment by thaJeztah** [comment]
I'm wondering though; if 2 images are tags for the same digest, wouldn't we have to use the reverse? Untag the image, and let containerd prune / garbage collect unused content?

(If we use "force", them probably we must change the tag name to a dangling image name instead)

**Comment by thaJeztah** [comment]
Discussing with @neersighted and we were considering if we should take a different approach in general;

- whenever we create an image, always create two images:
    - 1 image with the name:tag (if any)
    - 1 "dangling" image
- the second one is effectively replacing the graph driver "content addressable" store (access image by digest)


Now when we remove an image; we either remove the image's `name:tag` from containerd's store, or the dangling image name (if there's no more references to it).

So instead of creating dangling image references "on the fly", we can create them up-front, and use them for reference counting on our side

**Comment by vvoland** [comment]
Ah, the failure is on Windows only, because the `hack/make/.build-empty-images` doesn't run on Windows, so there's no special dangling image.

> So instead of creating dangling image references "on the fly", we can create them up-front, and use them for reference counting on our side.

I thought about it before, but it seemed to be a bit more tricky because the image creation is not always on our side (for example building image with Buildkit) whereas it's more common for Moby to perform image deletion.

On the other hand, that would indeed give us more control over the GC and would make more sense in cases where an explicit image digest is used instead of a name (so for example `docker run 3abd238` wouldn't pick **any** image that points to the `sha256:3abd238...`).

We could give it a try and see if that doesn't complicate things too much. It's a bit orthogonal to this PR though, so possibly we should just fix this issue here and then continue in a follow-up/separate ticket.

> I'm wondering though; if 2 images are tags for the same digest, wouldn't we have to use the reverse? Untag the image, and let containerd prune / garbage collect unused content?

Yes, this will cause the unused tag which also points to the same target as the used tag.
Looks like this can be fixed easily by just always excluding the dangling image of the container image.

Also, there's another unsolved issue -  images referenced by their id (`docker run 2376a0c12759aa` or `docker run sha256:...`) will be pruned. 

Fixed it now.

File: daemon/containerd/image_prune.go [hunk 1]
```diff
@@ -107,12 +125,25 @@ func (i *ImageService) pruneUnused(ctx context.Context, filterFunc imageFilterFu
...
+		totalCount := imagesWithDigest[dgst]
+		if totalCount == 1 {
+			if _, isUsed := usedDigests[dgst]; isUsed {
+				if err := i.ensureDanglingImage(ctx, img); err != nil {
+					return &report, errors.Wrapf(err, "failed to create ensure dangling image for %s", img.Name)
+				}
+			}
+		}
+		imagesWithDigest[dgst] = totalCount - 1
 	}
```
[the remark this answers is not in the document]
Comment (thaJeztah) [reply daemon/containerd/image_prune.go]: Somewhat wondering (but need to revisit the logic again) if we  could capture all of these in a single struct; something like;

```go
func prune() {
	// references to this digest, and whether they should be pruned.
	imageRefs := map[digest.Digest]map[string]bool{}

	imageRefs["sha256:deadbeef"] = map[string]bool{}

	// hello-world:latest can be pruned
	imageRefs["sha256:deadbeef"]["docker.io/library/hello-world:latest"] = true

	// or not ... something is still referencing it
	imageRefs["sha256:deadbeef"]["docker.io/library/hello-world:latest"] = false

	// container using the image, but by digest
	imageRefs["sha256:deadbeef"][danglingImageName("sha256:deadbeef")] = false

	for dgst, refs := range imageRefs {
		for ref, prune := range refs {
			if prune {
				// DELETE IMAGE
				// remove this reference from the digest
				delete(imageRefs[dgst], ref)
			}
		}
		if len(imageRefs[dgst]) == 0 {
			// no more references to this digest: DELETE CONTENT
		}
	}
}
```

File: daemon/containerd/image_prune.go [hunk 2]
```diff
@@ -107,11 +125,27 @@ func (i *ImageService) pruneUnused(ctx context.Context, filterFunc imageFilterFu
...
+
+		totalCount := imagesWithDigest[dgst]
+		if totalCount == 1 {
+			if _, isUsed := usedDigests[dgst]; isUsed {
+				if err := i.ensureDanglingImage(ctx, img); err != nil {
+					return &report, errors.Wrapf(err, "failed to create ensure dangling image for %s", img.Name)
+				}
+			}
+		}
+		imagesWithDigest[dgst] = totalCount - 1
```
Comment (neersighted) [inline daemon/containerd/image_prune.go:146]:
```suggestion
		if imagesWithDigest[dgst] == 1 {
			if _, isUsed := usedDigests[dgst]; isUsed {
				if err := i.ensureDanglingImage(ctx, img); err != nil {
					return &report, errors.Wrapf(err, "failed to create ensure dangling image for %s", img.Name)
				}
			}
		}
		imagesWithDigest[dgst] -= 1
```

**Review by neersighted — CHANGES_REQUESTED** [review]
Some initial nits, though I'm not a fan of how imperative this code is, and I think I might send a PR to your branch with a restructured implementation that keeps each category of prune exclusion more discrete.

File: integration/image/inspect_test.go [hunk 3]
```diff
@@ -19,10 +19,7 @@ func TestImageInspectEmptyTagsAndDigests(t *testing.T) {
 	client := testEnv.APIClient()
 	ctx := context.Background()
 
-	danglingId := environment.DanglingImageIdGraphDriver
-	if testEnv.UsingSnapshotter() {
-		danglingId = environment.DanglingImageIdSnapshotter
-	}
+	danglingId := environment.GetTestDanglingImageId(testEnv)
```
[the remark this answers is not in the document]
Comment (vvoland) [reply integration/image/inspect_test.go:22]: Changed, it's a small price for making your OCD side happy 😄

File: integration/image/prune_test.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+	danglingId := environment.GetTestDanglingImageId(testEnv)
+
+	cid := container.Create(ctx, t, client,
+		container.WithImage(danglingId),
+		container.WithCmd("sleep", "60"))
+
+	t.Cleanup(func() {
+		client.ContainerRemove(ctx, cid, types.ContainerRemoveOptions{Force: true})
+	})
```
[the remark this answers is not in the document]
Comment (vvoland) [reply integration/image/prune_test.go:32]: Removed 👍🏻

File: integration/image/prune_test.go [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+	"github.com/docker/docker/integration/internal/container"
+	"github.com/docker/docker/testutil/environment"
+	"gotest.tools/v3/assert"
+	is "gotest.tools/v3/assert/cmp"
+	"gotest.tools/v3/skip"
+)
+
+// Regression test for: https://github.com/moby/moby/issues/45732
+func TestPruneDontDeleteUsedDangling(t *testing.T) {
+	skip.If(t, testEnv.DaemonInfo.OSType == "windows")
```
[the remark this answers is not in the document]
Comment (vvoland) [reply integration/image/prune_test.go:18]: Done.

File: daemon/containerd/image_prune.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (vvoland) [reply daemon/containerd/image_prune.go]: Changed to decrement first.

File: daemon/containerd/image_prune.go [hunk 6]
```diff
@@ -70,35 +71,52 @@ func (i *ImageService) pruneUnused(ctx context.Context, filterFunc imageFilterFu
...
+	imagesWithDigest := map[digest.Digest]int{}
+	// Images considered for pruning.
 	imagesToPrune := map[string]containerdimages.Image{}
 	for _, img := range allImages {
-		if !danglingOnly || isDanglingImage(img) {
-			imagesToPrune[img.Name] = img
-		}
-	}
+		count := imagesWithDigest[img.Target.Digest]
+		imagesWithDigest[img.Target.Digest] = count + 1
```
[the remark this answers is not in the document]
Comment (vvoland) [reply daemon/containerd/image_prune.go:80]: Done.

File: daemon/containerd/image_prune.go [hunk 7]
```diff
@@ -70,35 +71,52 @@ func (i *ImageService) pruneUnused(ctx context.Context, filterFunc imageFilterFu
 		return nil, err
 	}
 
+	// How many images there are with each digest.
+	imagesWithDigest := map[digest.Digest]int{}
```
[the remark this answers is not in the document]
Comment (vvoland) [reply daemon/containerd/image_prune.go:75]: Done.

File: daemon/containerd/image_prune.go [hunk 8]
```diff
@@ -70,35 +71,52 @@ func (i *ImageService) pruneUnused(ctx context.Context, filterFunc imageFilterFu
...
-
 	var errs error
+
+	// Image specified by digests that are used by containers.
+	usedDigests := map[digest.Digest]struct{}{}
+
 	// Exclude images used by existing containers
-	for _, ctr := range containers {
+	for _, ctr := range i.containers.List() {
 		// Config.Image is the image reference passed by user.
```
[the remark this answers is not in the document]
Comment (vvoland) [reply daemon/containerd/image_prune.go:103]: Done.

File: daemon/containerd/image_prune.go [hunk 9]
```diff
@@ -70,35 +71,52 @@ func (i *ImageService) pruneUnused(ctx context.Context, filterFunc imageFilterFu
...
+		truncatedId := strings.HasPrefix(imageDgst.String(), "sha256:"+strings.TrimPrefix(ctr.Config.Image, "sha256:"))
+
+		// If the original image was deleted, make sure we don't delete the dangling image
+		delete(imagesToPrune, danglingImageName(ctr.ImageID.Digest()))
+
+		if truncatedId {
+			// If user didn't specify an explicit image, mark the digest as used.
+			usedDigests[imageDgst] = struct{}{}
+			continue
+		}
```
[the remark this answers is not in the document]
Comment (vvoland) [reply daemon/containerd/image_prune.go:118]: Changed to `normalizedImageID`.

File: daemon/containerd/image_prune.go [hunk 10]
```diff
@@ -70,35 +71,50 @@ func (i *ImageService) pruneUnused(ctx context.Context, filterFunc imageFilterFu
...
+		// Config.ImageID is the resolved content digest based on the user's Config.Image.
+		// For example: container created by:
+		//           `docker run alpine` will have Config.Image="alpine"
+		//           `docker run 82d1e9d` will have Config.Image="82d1e9d"
+		// but both will have ImageID="sha256:82d1e9d7ed48a7523bdebc18cf6290bdb97b82302a8a9c27d4fe885949ea94d1"
+		imageDgst := ctr.ImageID.Digest()
+
+		// If user didn't specify an explicit image, mark the digest as used.
+		normalizedImageID := strings.HasPrefix(imageDgst.String(), "sha256:"+strings.TrimPrefix(ctr.Config.Image, "sha256:"))
+		if normalizedImageID {
```
Comment (neersighted) [inline daemon/containerd/image_prune.go:113]: This still doesn't quite make sense -- the intention was to split the normalization from the check to make the intention clear:
```suggestion
		normalizedImageID := "sha256:"+strings.TrimPrefix(ctr.Config.Image, "sha256:")
		if strings.HasPrefix(imageDgst.String(), normalizedImageID) {
```

**Review by neersighted — APPROVED** [review]
One nit, but overall LGTM. I think that we should restructure this logic in the future, but I think that we can take this as an improvement for now, and think about lifecycle management more holistically in the future (during which we can revisit this logic, and perhaps even generalize it).

File: daemon/containerd/image_prune.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (thaJeztah) [reply daemon/containerd/image_prune.go]: I'll "unresolved" this one for future reference (if we're gonna look at possibly restructuring things)

File: daemon/containerd/image_prune.go (hunk 10 above)
Comment (vvoland) [reply daemon/containerd/image_prune.go:113]: Oh, sorry! I read the comment diff too quickly and assumed the wrong thing 😅 
Fixed now!

File: integration/image/inspect_test.go [hunk 11]
```diff
@@ -19,9 +19,9 @@ func TestImageInspectEmptyTagsAndDigests(t *testing.T) {
 	client := testEnv.APIClient()
 	ctx := context.Background()
 
-	danglingId := environment.GetTestDanglingImageId(testEnv)
+	danglingID := environment.GetTestDanglingImageId(testEnv)
```
Comment (thaJeztah) [inline integration/image/inspect_test.go:22]: Was this meant to go in the first commit? 🙈 

```diff
	danglingID := environment.GetTestDanglingImageID(testEnv)
```

**Review by thaJeztah — APPROVED** [review]
LGTM (but let me know if you want to address that nit; not a blocker)

File: integration/image/inspect_test.go (hunk 11 above)
Comment (vvoland) [reply integration/image/inspect_test.go:22]: 🙈  Fixing it up!

Comment (vvoland) [reply integration/image/inspect_test.go:22]: Done, I mistakenly did a `--fixup` of both tests to the same commit 😅

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by neersighted** [comment]
Backported at https://github.com/moby/moby/pull/45857

**Comment by thaJeztah** [comment]
@neersighted commits still align? I see there may be a potential race condition between you cherry-picking and Pawel force-pushing 😓 🙈

**Comment by neersighted** [comment]
That's one of the things to check when reviewing a backport PR 😉 

But yes, I made sure to re-cherry (with rerere) once Pawel pushed again; all SHAs are aligned to the latest version of this PR.

**Comment by thaJeztah** [comment]
Thanks!

**Comment by thaJeztah** [comment]
Timeout in one of the BuildKit tests (panic: test timed out after 30m0s); kicked it again

**Pull request merged**